File: pkg/ent/schema/propertycategory.go
// Copyright (c) 2004-present Facebook All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package schema

import (
	"github.com/facebook/ent"
	"github.com/facebook/ent/schema/edge"
	"github.com/facebook/ent/schema/field"
	"github.com/facebook/ent/schema/index"
)

// PropertyCategory defines the property type schema.
type PropertyCategory struct {
	schema
}

// Fields returns property type fields.
func (PropertyCategory) Fields() []ent.Field {
	return []ent.Field{
		field.String("name").
			NotEmpty().
			Unique(),
		field.Int("index"),
	}
}

// Edges returns property type edges.
func (PropertyCategory) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("properties_type", PropertyType.Type),
		edge.From("parameter_catalog", ParameterCatalog.Type).
			Ref("property_categories").
			Unique(),
	}
}

// Indexes returns property type indexes.
func (PropertyCategory) Indexes() []ent.Index {
	return []ent.Index{
		index.Fields("name"),
	}
}

File: app/fbcnms-projects/inventory/app/common/PropertyCategory.js
/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

import type {NamedNode, OptionalNamedNode} from './EntUtils';
import type {
  PropertyCategoryNodesQuery,
  PropertyCategoryNodesQueryVariables,
} from './__generated__/PropertyCategoryNodesQuery.graphql';

import {graphql} from './RelayUtils';
import {useLazyLoadQuery} from 'react-relay/hooks';

export type PropertyCategory = {|
  ...NamedNode,
  index?: ?number,
  numberOfProperties?: ?number,
|};

export const GENERAL_CATEGORY_LABEL = 'GENERAL';

const propertyCategoryNodesQuery = graphql`
  query PropertyCategoryNodesQuery($orderBy: PropertyCategoryOrder) {
    propertyCategories(orderBy: $orderBy) {
      edges {
        node {
          id
          name
        }
      }
    }
  }
`;

export type PropertyCategoryNode = $Exact<OptionalNamedNode>;

export function usePropertyCategoryNodes(
  input: PropertyCategoryNodesQueryVariables,
): $ReadOnlyArray<PropertyCategoryNode> {
  const response = useLazyLoadQuery<PropertyCategoryNodesQuery>(
    propertyCategoryNodesQuery,
    input,
  );
  const propCategoryData = response.propertyCategories?.edges || [];
  return propCategoryData.map(p => p.node).filter(Boolean);
}
